File: uploaddemo/src/business/galary/java/com/jw/galary/img/util/BitmapUtil.java
//
// Source code recreated from a .class file by IntelliJ IDEA
// (powered by Fernflower decompiler)
//

package com.jw.galary.img.util;

import android.app.Activity;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Matrix;
import android.graphics.Bitmap.CompressFormat;
import android.graphics.BitmapFactory.Options;
import android.media.ExifInterface;
import android.net.Uri;
import android.provider.MediaStore.Images.Media;
import android.text.TextUtils;
import android.util.Base64;
import java.io.BufferedInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;

public class BitmapUtil {
    private BitmapUtil() {
        throw new UnsupportedOperationException("u can't instantiate me...");
    }

    public static int getBitmapDegree(String path) {
        short degree = 0;

        try {
            ExifInterface exifInterface = new ExifInterface(path);
            int orientation = exifInterface.getAttributeInt("Orientation", 1);
            switch(orientation) {
                case 3:
                    degree = 180;
                    break;
                case 6:
                    degree = 90;
                    break;
                case 8:
                    degree = 270;
            }
        } catch (IOException var4) {
            var4.printStackTrace();
        }

        return degree;
    }

    public static Bitmap rotateBitmapByDegree(Bitmap bitmap, int degree) {
        Matrix matrix = new Matrix();
        matrix.postRotate((float)degree);
        Bitmap newBitmap = Bitmap.createBitmap(bitmap, 0, 0, bitmap.getWidth(), bitmap.getHeight(), matrix, true);
        if (!bitmap.isRecycled()) {
            bitmap.recycle();
        }

        return newBitmap;
    }

    public static Uri getRotatedUri(Activity activity, String path) {
        int degree = getBitmapDegree(path);
        if (degree != 0) {
            Bitmap bitmap = BitmapFactory.decodeFile(path);
            Bitmap newBitmap = rotateBitmapByDegree(bitmap, degree);
            return Uri.parse(Media.insertImage(activity.getContentResolver(), newBitmap, (String)null, (String)null));
        } else {
            return Uri.fromFile(new File(path));
        }
    }

    public static Bitmap rotateBitmapByDegree(String path, int degree) {
        Bitmap bitmap = BitmapFactory.decodeFile(path);
        return rotateBitmapByDegree(bitmap, degree);
    }

    public static Bitmap getBitmap(File file) {
        if (file == null) {
            return null;
        } else {
            BufferedInputStream is = null;

            Object var3;
            try {
                is = new BufferedInputStream(new FileInputStream(file));
                Bitmap var2 = BitmapFactory.decodeStream(is);
                return var2;
            } catch (FileNotFoundException var13) {
                var13.printStackTrace();
                var3 = null;
            } finally {
                try {
                    if (is != null) {
                        is.close();
                    }
                } catch (IOException var12) {
                    var12.printStackTrace();
                }

            }

            return (Bitmap)var3;
        }
    }

    public static Bitmap getBitmap(String filePath) {
        return TextUtils.isEmpty(filePath) ? null : BitmapFactory.decodeFile(filePath);
    }

    public static Bitmap getScaledBitmap(String filePath, int width, int height) {
        Options decodeOptions = new Options();
        decodeOptions.inJustDecodeBounds = true;
        BitmapFactory.decodeFile(filePath, decodeOptions);
        int actualWidth = decodeOptions.outWidth;
        int actualHeight = decodeOptions.outHeight;
        int maxWidth;
        int maxHeight;
        if (actualHeight < actualWidth) {
            maxWidth = height;
            maxHeight = width;
        } else {
            maxWidth = width;
            maxHeight = height;
        }

        int desiredWidth = getResizedDimension(maxWidth, maxHeight, actualWidth, actualHeight);
        int desiredHeight = getResizedDimension(maxHeight, maxWidth, actualHeight, actualWidth);
        decodeOptions.inJustDecodeBounds = false;
        decodeOptions.inSampleSize = findBestSampleSize(actualWidth, actualHeight, desiredWidth, desiredHeight);
        Bitmap tempBitmap = BitmapFactory.decodeFile(filePath, decodeOptions);
        Bitmap bitmap;
        if (tempBitmap == null || tempBitmap.getWidth() <= desiredWidth && tempBitmap.getHeight() <= desiredHeight) {
            bitmap = tempBitmap;
        } else {
            bitmap = Bitmap.createScaledBitmap(tempBitmap, desiredWidth, desiredHeight, true);
            tempBitmap.recycle();
        }

        return bitmap;
    }

    public static Bitmap getScaledBitmap(Context context, int imageResId, int maxWidth, int maxHeight) {
        Options decodeOptions = new Options();
        decodeOptions.inJustDecodeBounds = true;
        BitmapFactory.decodeResource(context.getResources(), imageResId, decodeOptions);
        int actualWidth = decodeOptions.outWidth;
        int actualHeight = decodeOptions.outHeight;
        int desiredWidth = getResizedDimension(maxWidth, maxHeight, actualWidth, actualHeight);
        int desiredHeight = getResizedDimension(maxHeight, maxWidth, actualHeight, actualWidth);
        decodeOptions.inJustDecodeBounds = false;
        decodeOptions.inSampleSize = findBestSampleSize(actualWidth, actualHeight, desiredWidth, desiredHeight);
        Bitmap tempBitmap = BitmapFactory.decodeResource(context.getResources(), imageResId, decodeOptions);
        Bitmap bitmap;
        if (tempBitmap == null || tempBitmap.getWidth() <= desiredWidth && tempBitmap.getHeight() <= desiredHeight) {
            bitmap = tempBitmap;
        } else {
            bitmap = Bitmap.createScaledBitmap(tempBitmap, desiredWidth, desiredHeight, true);
            tempBitmap.recycle();
        }

        return bitmap;
    }

    private static int findBestSampleSize(int actualWidth, int actualHeight, int desiredWidth, int desiredHeight) {
        double wr = (double)actualWidth / (double)desiredWidth;
        double hr = (double)actualHeight / (double)desiredHeight;
        double ratio = Math.min(wr, hr);

        float n;
        for(n = 1.0F; (double)(n * 2.0F) <= ratio; n *= 2.0F) {
        }

        return (int)n;
    }

    private static int getResizedDimension(int maxPrimary, int maxSecondary, int actualPrimary, int actualSecondary) {
        if (maxPrimary == 0 && maxSecondary == 0) {
            return actualPrimary;
        } else {
            double ratio;
            if (maxPrimary == 0) {
                ratio = (double)maxSecondary / (double)actualSecondary;
                return (int)((double)actualPrimary * ratio);
            } else if (maxSecondary == 0) {
                return maxPrimary;
            } else {
                ratio = (double)actualSecondary / (double)actualPrimary;
                int resized = maxPrimary;
                if ((double)maxPrimary * ratio > (double)maxSecondary) {
                    resized = (int)((double)maxSecondary / ratio);
                }

                return resized;
            }
        }
    }

    public static Bitmap compress(String srcImg) {
        Options options = new Options();
        options.inSampleSize = computeSize(options.outWidth, options.outHeight);
        return BitmapFactory.decodeFile(srcImg, options);
    }

    public static boolean saveBitmap2File(Bitmap bm, String savePath) {
        boolean saved = true;

        try {
            File f = new File(savePath);
            if (f.exists()) {
                f.delete();
            }

            FileOutputStream out = new FileOutputStream(f);
            bm.compress(CompressFormat.JPEG, 90, out);
            out.flush();
            out.close();
        } catch (FileNotFoundException var5) {
            var5.printStackTrace();
            saved = false;
        } catch (IOException var6) {
            var6.printStackTrace();
            saved = false;
        }

        return saved;
    }

    private static int computeSize(int srcWidth, int srcHeight) {
        srcWidth = srcWidth % 2 == 1 ? srcWidth + 1 : srcWidth;
        srcHeight = srcHeight % 2 == 1 ? srcHeight + 1 : srcHeight;
        int longSide = Math.max(srcWidth, srcHeight);
        int shortSide = Math.min(srcWidth, srcHeight);
        float scale = (float)shortSide / (float)longSide;
        if (scale <= 1.0F && (double)scale > 0.5625D) {
            if (longSide < 1664) {
                return 1;
            } else if (longSide >= 1664 && longSide < 4990) {
                return 2;
            } else if (longSide > 4990 && longSide < 10240) {
                return 4;
            } else {
                return longSide / 1280 == 0 ? 1 : longSide / 1280;
            }
        } else if ((double)scale <= 0.5625D && (double)scale > 0.5D) {
            return longSide / 1280 == 0 ? 1 : longSide / 1280;
        } else {
            return (int)Math.ceil((double)longSide / (1280.0D / (double)scale));
        }
    }

    public static String base64Image(String filePath) {
        String imgBase64 = "";

        try {
            File file = new File(filePath);
            byte[] content = new byte[(int)file.length()];
            FileInputStream finputstream = new FileInputStream(file);
            finputstream.read(content);
            finputstream.close();
            imgBase64 = Base64.encodeToString(content, 0);
        } catch (IOException var5) {
            var5.printStackTrace();
        }

        return imgBase64;
    }
}

File: library/src/main/java/com/jw/library/utils/BitmapUtil.kt
package com.jw.library.utils

import android.app.Activity
import android.content.Context
import android.graphics.Bitmap
import android.graphics.Bitmap.CompressFormat
import android.graphics.BitmapFactory
import android.graphics.BitmapFactory.Options
import android.graphics.Matrix
import android.media.ExifInterface
import android.net.Uri
import android.provider.MediaStore
import android.provider.MediaStore.Images.Media
import android.text.TextUtils
import android.util.Base64
import java.io.*

/**
 * Bitmap工具类
 */
object BitmapUtil {


    /**
     * 根据路径获取图片旋转角度
     * @param path String?
     * @return Int
     */
    fun getBitmapDegree(path: String?): Int {
        var degree: Short = 0

        try {
            val exifInterface = ExifInterface(path)
            val orientation = exifInterface.getAttributeInt("Orientation", 1)
            when (orientation) {
                3 -> degree = 180
                6 -> degree = 90
                8 -> degree = 270
            }
        } catch (var4: IOException) {
            var4.printStackTrace()
        }

        return degree.toInt()
    }

    /**
     * 旋转图片
     * @param bitmap Bitmap?
     * @param degree Int
     * @return Bitmap
     */
    fun rotateBitmapByDegree(bitmap: Bitmap?, degree: Int): Bitmap {
        val matrix = Matrix()
        matrix.postRotate(degree.toFloat())
        val newBitmap =
            Bitmap.createBitmap(bitmap!!, 0, 0, bitmap.width, bitmap.height, matrix, true)
        if (!bitmap.isRecycled) {
            bitmap.recycle()
        }

        return newBitmap
    }

    /**
     * 旋转图片
     * @param context Context
     * @param uri Uri
     * @param options Options
     * @return Bitmap?
     * @throws FileNotFoundException
     */
    @Throws(FileNotFoundException::class)
    fun rotateBitmap(context: Context, uri: Uri, options: Options): Bitmap? {
        val bis = context.contentResolver.openInputStream(uri)

        try {
            var result = BitmapFactory.decodeStream(bis, null, options)
            val degree = getBitmapDegree(uri.path)
            if (degree != 0) {
                result = rotateBitmapByDegree(result, degree)
            }
            return result
        } catch (var7: OutOfMemoryError) {
            if (options.inSampleSize >= 64) {
                return null
            }

            options.inSampleSize *= 2
            return null
        }

    }

    /**
     * 旋转图片
     * @param path String
     * @param degree Int
     * @return Bitmap
     */
    fun rotateBitmapByDegree(path: String, degree: Int): Bitmap {
        val bitmap = BitmapFactory.decodeFile(path)
        if (degree == 0)
            return bitmap
        return rotateBitmapByDegree(bitmap, degree)
    }

    /**
     * 获取正常方向的图片
     * @param activity Activity
     * @param path String
     * @return Uri
     */
    fun getRotatedUri(activity: Activity, path: String): Uri {
        val degree = getBitmapDegree(path)
        if (degree != 0) {
            val bitmap = BitmapFactory.decodeFile(path)
            val newBitmap = rotateBitmapByDegree(bitmap, degree)
            return Uri.parse(Media.insertImage(activity.contentResolver, newBitmap, null, null))
        } else {
            return Uri.fromFile(File(path))
        }
    }

    /**
     * 从文件中获取bitmap
     * @param file File?
     * @return Bitmap?
     */
    fun getBitmap(file: File?): Bitmap? {
        if (file == null) {
            return null
        } else {
            var bis: BufferedInputStream? = null

            val var3: Any?
            try {
                bis = BufferedInputStream(FileInputStream(file))
                return BitmapFactory.decodeStream(bis)
            } catch (var13: FileNotFoundException) {
                var13.printStackTrace()
                var3 = null
            } finally {
                try {
                    bis?.close()
                } catch (var12: IOException) {
                    var12.printStackTrace()
                }

            }

            return var3 as Bitmap?
        }
    }

    /**
     * 从路径中获取bitmap
     * @param filePath String
     * @return Bitmap?
     */
    fun getBitmap(filePath: String): Bitmap? {
        return if (TextUtils.isEmpty(filePath)) null else BitmapFactory.decodeFile(filePath)
    }

    /**
     * 根据宽度缩放bitmap
     * @param filePath String
     * @param width Int
     * @param height Int
     * @return Bitmap?
     */
    fun getScaledBitmap(filePath: String, width: Int, height: Int): Bitmap? {
        val decodeOptions = Options()
        decodeOptions.inJustDecodeBounds = true
        BitmapFactory.decodeFile(filePath, decodeOptions)
        val actualWidth = decodeOptions.outWidth
        val actualHeight = decodeOptions.outHeight
        val maxWidth: Int
        val maxHeight: Int
        if (actualHeight < actualWidth) {
            maxWidth = height
            maxHeight = width
        } else {
            maxWidth = width
            maxHeight = height
        }

        val desiredWidth = getResizedDimension(maxWidth, maxHeight, actualWidth, actualHeight)
        val desiredHeight = getResizedDimension(maxHeight, maxWidth, actualHeight, actualWidth)
        decodeOptions.inJustDecodeBounds = false
        decodeOptions.inSampleSize =
            findBestSampleSize(actualWidth, actualHeight, desiredWidth, desiredHeight)
        val tempBitmap = BitmapFactory.decodeFile(filePath, decodeOptions)
        val bitmap: Bitmap?
        if (tempBitmap == null || tempBitmap.width <= desiredWidth && tempBitmap.height <= desiredHeight) {
            bitmap = tempBitmap
        } else {
            bitmap = Bitmap.createScaledBitmap(tempBitmap, desiredWidth, desiredHeight, true)
            tempBitmap.recycle()
        }

        return bitmap
    }

    /**
     * 获取指定宽高bitmap
     * @param context Context
     * @param imageResId Int
     * @param maxWidth Int
     * @param maxHeight Int
     * @return Bitmap?
     */
    fun getScaledBitmap(
        context: Context,
        imageResId: Int,
        maxWidth: Int,
        maxHeight: Int
    ): Bitmap? {
        val decodeOptions = Options()
        decodeOptions.inJustDecodeBounds = true
        BitmapFactory.decodeResource(context.resources, imageResId, decodeOptions)
        val actualWidth = decodeOptions.outWidth
        val actualHeight = decodeOptions.outHeight
        val desiredWidth = getResizedDimension(maxWidth, maxHeight, actualWidth, actualHeight)
        val desiredHeight = getResizedDimension(maxHeight, maxWidth, actualHeight, actualWidth)
        decodeOptions.inJustDecodeBounds = false
        decodeOptions.inSampleSize =
            findBestSampleSize(actualWidth, actualHeight, desiredWidth, desiredHeight)
        val tempBitmap =
            BitmapFactory.decodeResource(context.resources, imageResId, decodeOptions)
        val bitmap: Bitmap?
        if (tempBitmap == null || tempBitmap.width <= desiredWidth && tempBitmap.height <= desiredHeight) {
            bitmap = tempBitmap
        } else {
            bitmap = Bitmap.createScaledBitmap(tempBitmap, desiredWidth, desiredHeight, true)
            tempBitmap.recycle()
        }

        return bitmap
    }

    private fun findBestSampleSize(
        actualWidth: Int,
        actualHeight: Int,
        desiredWidth: Int,
        desiredHeight: Int
    ): Int {
        val wr = actualWidth.toDouble() / desiredWidth.toDouble()
        val hr = actualHeight.toDouble() / desiredHeight.toDouble()
        val ratio = Math.min(wr, hr)

        var n: Float
        n = 1.0f
        while ((n * 2.0f).toDouble() <= ratio) {
            n *= 2.0f
        }

        return n.toInt()
    }

    private fun getResizedDimension(
        maxPrimary: Int,
        maxSecondary: Int,
        actualPrimary: Int,
        actualSecondary: Int
    ): Int {
        if (maxPrimary == 0 && maxSecondary == 0) {
            return actualPrimary
        } else {
            val ratio: Double
            if (maxPrimary == 0) {
                ratio = maxSecondary.toDouble() / actualSecondary.toDouble()
                return (actualPrimary.toDouble() * ratio).toInt()
            } else if (maxSecondary == 0) {
                return maxPrimary
            } else {
                ratio = actualSecondary.toDouble() / actualPrimary.toDouble()
                var resized = maxPrimary
                if (maxPrimary.toDouble() * ratio > maxSecondary.toDouble()) {
                    resized = (maxSecondary.toDouble() / ratio).toInt()
                }

                return resized
            }
        }
    }

    fun compress(srcImg: String): Bitmap {
        val options = Options()
        options.inSampleSize = computeSize(options.outWidth, options.outHeight)
        return BitmapFactory.decodeFile(srcImg, options)
    }

    /**
     * 将bitmap保存为文件
     * @param bm Bitmap
     * @param savePath String   保存路径
     * @return Boolean
     */
    fun saveBitmap2File(bm: Bitmap, savePath: String): Boolean {
        var saved = true

        try {
            val f = File(savePath)
            if (f.exists()) {
                f.delete()
            }

            val out = FileOutputStream(f)
            bm.compress(CompressFormat.JPEG, 90, out)
            out.flush()
            out.close()
        } catch (var5: FileNotFoundException) {
            var5.printStackTrace()
            saved = false
        } catch (var6: IOException) {
            var6.printStackTrace()
            saved = false
        }

        return saved
    }

    private fun computeSize(srcWidth: Int, srcHeight: Int): Int {
        var srcWidth = srcWidth
        var srcHeight = srcHeight
        srcWidth = if (srcWidth % 2 == 1) srcWidth + 1 else srcWidth
        srcHeight = if (srcHeight % 2 == 1) srcHeight + 1 else srcHeight
        val longSide = Math.max(srcWidth, srcHeight)
        val shortSide = Math.min(srcWidth, srcHeight)
        val scale = shortSide.toFloat() / longSide.toFloat()
        return if (scale <= 1.0f && scale.toDouble() > 0.5625) {
            if (longSide < 1664) {
                1
            } else if (longSide >= 1664 && longSide < 4990) {
                2
            } else if (longSide > 4990 && longSide < 10240) {
                4
            } else {
                if (longSide / 1280 == 0) 1 else longSide / 1280
            }
        } else if (scale.toDouble() <= 0.5625 && scale.toDouble() > 0.5) {
            if (longSide / 1280 == 0) 1 else longSide / 1280
        } else {
            Math.ceil(longSide.toDouble() / (1280.0 / scale.toDouble())).toInt()
        }
    }

    /**
     * 将bitmap存入相册
     * @param bm Bitmap
     * @param activity Activity
     * @return Uri
     */
    fun saveBitmap2Galary(bm: Bitmap, activity: Activity): Uri {
        val path = MediaStore.Images.Media.insertImage(activity.contentResolver, bm, null, null)
        return Uri.parse(path)
    }

    /**
     * 将bitmap转为字节
     * @param bm Bitmap
     * @return ByteArray
     */
    fun Bitmap2Bytes(bm: Bitmap): ByteArray {
        val baos = ByteArrayOutputStream()
        bm.compress(CompressFormat.JPEG, 100, baos)
        return baos.toByteArray()
    }

    fun base64Image(filePath: String): String {
        var imgBase64 = ""

        try {
            val file = File(filePath)
            val content = ByteArray(file.length().toInt())
            val finputstream = FileInputStream(file)
            finputstream.read(content)
            finputstream.close()
            imgBase64 = Base64.encodeToString(content, 0)
        } catch (var5: IOException) {
            var5.printStackTrace()
        }

        return imgBase64
    }
}
